### PR TITLE
Preparation for RelationImpl. changes / Fix handling of combinations of relation types

### PR DESCRIPTION
**Pull request by CodeLionX**
## Proposed Changes

  - Refactor handling of binary relation operations (like `join`, `union`)
  - Fix missing handling of combination of different `Relation` subtypes for those bin. operations
  - Test for those combinations.

## Imperfections

- binary operations on `FutureRelation` just return type `Relation`, to work with the future relation, one has to cast it back to `FutureRelation`:
  ```scala
  val priceList: FutureRelation = Dactor.askDactor(
    system,
    classOf[StoreSection],
    priceRequests)
  val priceDisc: FutureRelation = priceList.innerJoin(fixedDiscount,
    (priceRec, discRec) =>
      priceRec.get(CartPurchases.inventoryId) == discRec.get(CartPurchases.inventoryId)
  ).asInstanceOf[FutureRelation]
  priceDisc.pipeAsMessageTo(...)
  ```
- One has to change `RelationBinOps`, if there will be new descendants of `Relation` and provide a `<NewRelationType>.BinOps`-implementation for it
- matching on different `Relation` subtypes is done in `RelationBinOps` as well as in the `BinOps` objects of the implementing types

### CODE REVIEW

**Comment by CodeLionX** [comment]
@srfc what do you think of the codacy integration's review-bot? Should I keep it turned on?

**Comment by CodeLionX** [comment]
@srfc Does this make sense for you?

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala [hunk 1]
```diff
@@ -97,31 +97,32 @@ object Cart {
...
-          CartPurchases.sectionId ~> rec.get(CartPurchases.sectionId).get &
-          CartPurchases.sessionId ~> rec.get(CartPurchases.sessionId).get &
-          CartPurchases.quantity ~> rec.get(CartPurchases.quantity).get &
-          CartPurchases.inventoryId ~> rec.get(StoreSection.Inventory.inventoryId).get &
-          CartPurchases.fixedDiscount ~> rec.get(GroupManager.Discounts.fixedDisc).get &
-          CartPurchases.minPrice ~> rec.get(StoreSection.Inventory.minPrice).get &
-          CartPurchases.price ~> rec.get(StoreSection.Inventory.price).get
-        ).build())
+      val result: FutureRelation = priceDiscOrder.transform( relation => Relation(
+        relation.records.get
```
[the remark this answers is not in the document]
Comment (srfc) [reply sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: The alternative seems like it is too verbose at no semantical gain (we want to fail anyway if the `Optional` is empty) to me.

File: sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala [hunk 2]
```diff
@@ -97,31 +97,32 @@ object Cart {
...
+          .map( (rec: Record) => rec + (CartPurchases.sessionId -> currentSessionId))
+          .map( (rec: Record) => CartPurchases.newRecord(
+            CartPurchases.sectionId ~> rec.get(CartPurchases.sectionId).get &
```
[the remark this answers is not in the document]
Comment (srfc) [reply sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala]: See above

**Pull request merged**